File: app/src/main/java/xyz/volgoak/wordlearning/recycler/RecyclerAdapter.java
package xyz.volgoak.wordlearning.recycler;

import android.content.Context;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.View;

import java.util.List;

import xyz.volgoak.wordlearning.entities.DataEntity;

/**
 * Created by alex on 1/3/18.
 */

public abstract class RecyclerAdapter<RC extends NewRowController> extends RecyclerView.Adapter<RC> {
    public static final String TAG = "CursorRecyclerAdapter";

    protected Context mContext;
    protected List<DataEntity> mEntities;
    private AdapterClickListener mAdapterClickListener;
    private AdapterLongClickListener mAdapterLongClickListener;
    private RecyclerView mRecyclerView;
    private boolean mSelectable;
    private ChoiceMode mChoiceMode;

    public RecyclerAdapter(Context context, List<DataEntity> entities, RecyclerView rv){
        mContext = context;
        mEntities = entities;
        mRecyclerView = rv;
    }

    @Override
    public void onBindViewHolder(NewRowController controller, int position) {
        controller.setSelectable(mSelectable);
        if(mChoiceMode != null){
            controller.setChecked(mChoiceMode.isChecked(position));
        }
        controller.bindController(mEntities.get(position));
    }

    @Override
    public int getItemCount() {
        return mEntities.size();
    }

    @Override
    public long getItemId(int position) {
        return mEntities.get(position).getId();
    }

    private void setSelectable(boolean selectable){
        if(mSelectable != selectable) {
            mSelectable = selectable;
            for (int a = 0; a < getItemCount(); a++) {
                NewRowController controller = (NewRowController) mRecyclerView.findViewHolderForAdapterPosition(a);

                if (controller != null) {
                    controller.setSelectable(selectable);
                    if(selectable && mChoiceMode != null){
                        controller.setChecked(mChoiceMode.isChecked(controller.getAdapterPosition()));
                    }
                }
            }

            //dirty trick
            LinearLayoutManager llm =(LinearLayoutManager) mRecyclerView.getLayoutManager();
            int firstVisible = llm.findFirstVisibleItemPosition();
            int lastVisible = llm.findLastVisibleItemPosition();

            for(int a = 1; a < 4 ; a++){
                notifyItemChanged(firstVisible - a);
                notifyItemChanged(lastVisible + a);
            }
        }
    }

    public void setChoiceMode(ChoiceMode choiceMode){
        mChoiceMode = choiceMode;
        if(choiceMode instanceof MultiChoiceMode){
            setSelectable(true);
        }else setSelectable(false);
    }

    public void changeData(List<DataEntity> entities){
        mEntities = entities;
        notifyDataSetChanged();
    }

    public List<DataEntity> getData(){
        return mEntities;
    }

    public Context getContext(){
        return mContext;
    }

    public void setAdapterClickListener(AdapterClickListener listener){
        mAdapterClickListener = listener;
    }

    public void setAdapterLongClickListener(AdapterLongClickListener adapterLongClickListener) {
        mAdapterLongClickListener = adapterLongClickListener;
    }

    public void onControllerClick(NewRowController controller, View root, int position){
        if(mChoiceMode != null){
            if(mChoiceMode instanceof SingleChoiceMode) {
                if(mChoiceMode.getCheckedPosition() != -1){
                    NewRowController rc = (NewRowController)
                            mRecyclerView.findViewHolderForAdapterPosition(mChoiceMode.getCheckedPosition());
                    if(rc != null) rc.setChecked(false);
                }
                controller.setChecked(true);
                mChoiceMode.setChecked(position, true);
            }else if(mChoiceMode instanceof MultiChoiceMode) {
                //change checked state
                boolean checked = !mChoiceMode.isChecked(position);
                controller.setChecked(checked);
                mChoiceMode.setChecked(position, checked);
            }
        }
        if(mAdapterClickListener != null){
            mAdapterClickListener.onClick(root, position, getItemId(position));
        }
    }

    public boolean onControllerLongClick(NewRowController controller, View root, int position){
        if(mAdapterLongClickListener != null){
            return  mAdapterLongClickListener.onLongClick(root, position, getItemId(position));
        }
        return false;
    }

    public void notifyEntityChanged(DataEntity dataEntity) {
        int position = mEntities.indexOf(dataEntity);
        if(position != -1) {
            mEntities.set(position, dataEntity);
            notifyItemChanged(position);
        }
    }

    public interface AdapterClickListener{
        void onClick(View root, int position, long id);
    }

    public interface AdapterLongClickListener {
        boolean onLongClick(View root, int position, long id);
    }
}

File: app/src/bottom_nav/java/xyz/volgoak/wordlearning/adapter/RecyclerAdapter.kt
package xyz.volgoak.wordlearning.adapter

import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.View

import xyz.volgoak.wordlearning.entities.DataEntity
import xyz.volgoak.wordlearning.recycler.ChoiceMode
import xyz.volgoak.wordlearning.recycler.MultiChoiceMode
import xyz.volgoak.wordlearning.recycler.SingleChoiceMode

/**
 * Created by alex on 1/3/18.
 */

abstract class RecyclerAdapter<RC : BaseHolder, DE: DataEntity>
(var entities: MutableList<DE>, private val mRecyclerView: RecyclerView) : RecyclerView.Adapter<RC>() {

    var onClick: (DE) -> Unit = {}
    var onLongClick: (DE) -> Unit = {}

    private var isSelectable: Boolean = false
    private var choiceMode: ChoiceMode? = null


    override fun onBindViewHolder(controller: RC, position: Int) {
        controller.setSelectable(isSelectable)
        if (choiceMode != null) {
            controller.setChecked(choiceMode!!.isChecked(position))
        }
        controller.bindController(entities[position])
    }

    override fun getItemCount(): Int {
        return entities.size
    }

    override fun getItemId(position: Int): Long {
        return entities[position].id
    }

    private fun setSelectable(selectable: Boolean) {
        if (isSelectable != selectable) {
            isSelectable = selectable
            for (a in 0 until itemCount) {
                val controller = mRecyclerView.findViewHolderForAdapterPosition(a) as BaseHolder?

                if (controller != null) {
                    controller.setSelectable(selectable)
                    if (selectable && choiceMode != null) {
                        controller.setChecked(choiceMode!!.isChecked(controller.adapterPosition))
                    }
                }
            }

            //dirty trick
            val llm = mRecyclerView.layoutManager as LinearLayoutManager
            val firstVisible = llm.findFirstVisibleItemPosition()
            val lastVisible = llm.findLastVisibleItemPosition()

            for (a in 1..3) {
                notifyItemChanged(firstVisible - a)
                notifyItemChanged(lastVisible + a)
            }
        }
    }

    fun setChoiceMode(choiceMode: ChoiceMode) {
        this.choiceMode = choiceMode
        if (choiceMode is MultiChoiceMode) {
            setSelectable(true)
        } else
            setSelectable(false)
    }

    fun changeData(entities: MutableList<DE>) {
        this.entities = entities
        notifyDataSetChanged()
    }

    open fun onControllerClick(controller: BaseHolder, root: View, position: Int) {
        if (choiceMode != null) {
            if (choiceMode is SingleChoiceMode) {
                if (choiceMode!!.checkedPosition != -1) {
                    val rc = mRecyclerView.findViewHolderForAdapterPosition(choiceMode!!.checkedPosition) as BaseHolder
                    rc.setChecked(false)
                }
                controller.setChecked(true)
                choiceMode!!.setChecked(position, true)
            } else if (choiceMode is MultiChoiceMode) {
                //change checked state
                val checked = !choiceMode!!.isChecked(position)
                controller.setChecked(checked)
                choiceMode!!.setChecked(position, checked)
            }
        }

        onClick(entities[position])
    }

    fun notifyEntityChanged(dataEntity: DE) {
        val position = entities.indexOf(dataEntity)
        if (position != -1) {
            entities[position] = dataEntity
            notifyItemChanged(position)
        }
    }
}
